src/Middleware/HeaderPropagation/src/HeaderPropagationMiddleware.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    /// <summary>
+    /// A Middleware for propagating headers to a <see cref="HttpClient"/>.
+    /// </summary>
+    public class HeaderPropagationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HeaderPropagationOptions _options;
+        private readonly HeaderPropagationValues _values;
+
+        public HeaderPropagationMiddleware(RequestDelegate next, IOptions<HeaderPropagationOptions> options, HeaderPropagationValues values)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _options = options.Value;
+
+            _values = values ?? throw new ArgumentNullException(nameof(values));
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            foreach ((var headerName, var entry) in _options.Headers)
+            {
+                var values = GetValues(headerName, entry, context);
+
+                if (!StringValues.IsNullOrEmpty(values))
+                {
+                    _values.Headers.TryAdd(headerName, values);
+                }
+            }
+
+            return _next.Invoke(context);
+        }
+
+        private static StringValues GetValues(string headerName, HeaderPropagationEntry entry, HttpContext context)
+        {
+            if (entry?.ValueFactory != null)
```

Comment:
What are the cases where entry could be null?

 Are you just hedging against someone adding a null entry to `_options.Headers`? I think it would be better to let this crash with an nullref.

Reply:
I think this could be a valid scenario, now that the header inbound name is the key of the dictionary and not in the entry.
A null configuration entry will correspond to using the same name outbound and having no factory nor default value, see last 2 tests: https://github.com/aspnet/AspNetCore/pull/7921/files#diff-e52877730c5607887b50f212d490c111R189

I guess I'll have to document this, unless you think an entry should be mandatory (eventually with all fields as null)

Reply:
I'm happy to leave this for now. When this gets merged I'm going to spend a little time playing with it.